analysis/parser/RawTextParser/src/main/kotlin/de/maibornwolff/codecharta/parser/rawtextparser/MetricCollector.kt

```diff
@@ -0,0 +1,60 @@
+package de.maibornwolff.codecharta.parser.rawtextparser
+
+import de.maibornwolff.codecharta.parser.rawtextparser.metrics.MetricsFactory
+import de.maibornwolff.codecharta.parser.rawtextparser.model.FileMetrics
+import java.io.File
+import java.nio.file.Paths
+
+class MetricCollector(private var root: File,
+                      private val exclude: Array<String> = arrayOf(),
+                      private val fileExtensions: Array<String> = arrayOf(),
+                      private val parameters: Map<String, Int> = mapOf(),
+                      private val metrics: List<String> = listOf()) {
+
+    val MAX_FILE_NAME_PRINT_LENGTH = 30
+
+    fun parse(): Map<String, FileMetrics> {
+        val projectMetrics = mutableMapOf<String, FileMetrics>()
+        val excludePatterns = exclude.joinToString(separator = "|", prefix = "(", postfix = ")").toRegex()
+
+        root.walk().forEach {
+            if (it.isFile) {
+                val standardizedPath = "/" + getRelativeFileName(it.toString())
+                val isMatchingFileExtension = fileExtensions.isEmpty() || fileExtensions.contains(standardizedPath.substringAfterLast("."))
+                val isNotExcluded = !(exclude.isNotEmpty() && excludePatterns.containsMatchIn(standardizedPath))
+                if (it.isFile && isNotExcluded && isMatchingFileExtension) {
+                    logProgress(it.name)
+                    projectMetrics[standardizedPath] = parseFile(it)
+                }
+            }
+        }
+        return projectMetrics
+    }
+
+    private fun parseFile(file: File): FileMetrics {
+        val metrics = MetricsFactory.create(metrics, parameters)
+        file.readLines().forEach { line -> metrics.forEach { it.parseLine(line) } }
```

Comment:
I don't think a metric should parse a line. Maybe a MetricVisitor can do that and then set a value in the actual metrics once it decided what the final metric for the file will be. If I remember correctly, sonar did it similar.

This basically means to split up the Metric-Interface into an actual Metric-Interface and a Metric-Visitor that supports methods such as `parseLine` or `setParameters`

Reply:
Personally I think it is unnecessary to split the Interface, since the resulting Metric classes will do nothing to very little. However, we can rename the metric classes to reflect better reflect what it is doing. Something like MetricCalculator or similar.

Or how would you have suggested to split the Interface